thrust/system/hip/hipstdpar/hipstdpar_lib.hpp

```diff
@@ -2278,15 +2278,301 @@
         // END NONE_OF
 
         // BEGIN NTH_ELEMENT
-        // TODO: UNIMPLEMENTED IN THRUST
+        template <typename KeysIt,
+                  typename CompareOp,
+                  enable_if_t<!hipstd::is_offloadable_iterator<KeysIt>()
+                              || !hipstd::is_offloadable_callable<CompareOp>()>* = nullptr>
+        inline void nth_element(execution::parallel_unsequenced_policy,
+                                KeysIt    first,
+                                KeysIt    nth,
+                                KeysIt    last,
+                                CompareOp compare_op)
+        {
+            if constexpr(!hipstd::is_offloadable_iterator<KeysIt>())
+            {
+                hipstd::unsupported_iterator_category<
+                    typename iterator_traits<KeysIt>::iterator_category>();
+            }
+            if constexpr(!hipstd::is_offloadable_callable<CompareOp>())
+            {
+                hipstd::unsupported_callable_type<CompareOp>();
+            }
+
+            std::nth_element(std::execution::par, first, nth, last, std::move(compare_op));
+        }
+
+        template <typename KeysIt,
+                  typename CompareOp,
+                  enable_if_t<hipstd::is_offloadable_iterator<KeysIt>()
+                              && hipstd::is_offloadable_callable<CompareOp>()>* = nullptr>
+        inline void nth_element(execution::parallel_unsequenced_policy,
+                                KeysIt    first,
+                                KeysIt    nth,
+                                KeysIt    last,
+                                CompareOp compare_op)
+        {
+            const size_t count = static_cast<size_t>(thrust::distance(first, last));
+            const size_t n     = static_cast<size_t>(thrust::distance(first, nth));
+
+            if(count == 0)
+            {
+                return;
+            }
+
+            auto        policy       = thrust::device;
+            size_t      storage_size = 0;
+            hipStream_t stream       = thrust::hip_rocprim::stream(policy);
+            bool        debug_sync   = THRUST_HIP_DEBUG_SYNC_FLAG;
+
+            hipError_t status;
+
+            status = rocprim::nth_element(
```

Comment:
The reason behind forwarding to rocPRIM's algorithms is that currently rocThrust does not have support for the new Parallel STL algorithms that have been added to rocPRIM, as neither NVIDIA's Thrust nor CUB have implemented those yet.

Reply:
It'd be somewhat less cognitively challenging to have something like a stub `thrust::__nth_element(...)` call (we'd have to add the stubs, but you're already splitting into impl headers, no?) with a `// TODO: this'll go away once something happens` rather than having inline prim for a few algorithms, whilst most others forward to the higher level thrust interface. You'll note that the originals tried to keep away from directly doing any `hip*` stuff, and trying to do only C++ level handling before forwarding to Thrust. I am concerned with some problematic future, where some algorithms just doing a bunch of rocprim stuff inline because why not, some others keeping the forward to Thrust approach and, possibly, a 3rd category doing something else:) Perhaps I'm the odd person out with this preference / concern, though.

Reply:
Sorry for the late answer. After discussing it a little bit internally, we are unsure whether adding stubs to rocThrust would be ideal: the interface of Thrust is _based_ on that of the C++ Standard Library, but is not guaranteed to be exactly the same, so adding new API entries to rocThrust without knowing how Thrust would do it means that we may need to change this interface later (when Thrust adds these algorithms), which would be a breaking change. 
If the only reason for adding stubs to rocThrust is removing the calls to rocPRIM from hipstdpar, then it's probably not worth the risk. Perhaps you have other worries about this that we are failing to see?

Also, about this:
> I am concerned with some problematic future, where some algorithms just doing a bunch of rocprim stuff inline because why not, some others keeping the forward to Thrust approach and, possibly, a 3rd category doing something else:)

I don't think it should be a concern, as the missing algorithms are exclusively being added to rocPRIM.

Reply:
I guess I don't really follow what the risk is. We're not subcontractors for the Thrust development team, so having additional interfaces wouldn't be a problem, if we were to go in that direction. Furthermore, perhaps the point is lost, but what I am suggesting is to do something along the lines of `thrust::__nth_element_placeholder_stub_do_not_use_danger_will_robinson`, where the latter holds the current rocprim calls. The definition for this can and should live in the implementation header. It's primarily a matter of symmetry and code consistency, no other algorithm calls Prim inline.

Reply:
Yup, that's it; then, when / if upstream adds these, we can just go and delete the double underscore prefix and the impl. This prevents folks (future maintainers) from having to ask themselves "should I just use PRIM here, or should I call Thrust", and at a glance makes it obvious we always call Thrust from the overloads. Thank you for taking the time to go through this, it's highly appreciated!

Reply:
Added the changes for this! Please check if it fits your idea for the use of stubs.